testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/wadl/WadlTestApplication.java

```diff
@@ -0,0 +1,100 @@
+package org.jboss.resteasy.test.wadl;
+
+import org.jboss.resteasy.utils.TestApplication;
+import org.jboss.resteasy.wadl.ResteasyWadlDefaultResource;
+import org.jboss.resteasy.wadl.ResteasyWadlWriter;
+
+import javax.ws.rs.ApplicationPath;
+import javax.ws.rs.core.Application;
+import javax.ws.rs.ext.Provider;
+import java.io.InputStream;
+import java.util.HashSet;
+import java.util.Set;
+
+@Provider
+@ApplicationPath("/")
+public class WadlTestApplication extends Application {
+   public static Set<Class<?>> classes = new HashSet<Class<?>>();
+   public static Set<Object> singletons = new HashSet<Object>();
+
+   /**
+    * Load resources from classes.txt file from deployment
+    *
+    * @return Array of class names.
+    */
+   public static String[] getClassesFromDeployment(String name) {
+      String resource = name + ".txt";
+      String stripped = resource.startsWith("/") ?
+            resource.substring(1) : resource;
+
+      InputStream stream = null;
+      ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+      if (classLoader != null) {
+         stream = classLoader.getResourceAsStream(stripped);
+      }
+      if (stream == null) {
+         stream = TestApplication.class.getResourceAsStream(resource);
+      }
+      if (stream == null) {
+         stream = TestApplication.class.getClassLoader().getResourceAsStream(stripped);
+      }
+      if (stream == null) {
+         return new String[0];
+      }
+      return convertStreamToString(stream).split(",");
+   }
+
+   /**
+    * Convert input stream to String
+    *
+    * @param is Input stream
+    * @return string
+    */
+   private static String convertStreamToString(final java.io.InputStream is) {
+      java.util.Scanner s = new java.util.Scanner(is).useDelimiter("\\A");
+      return s.hasNext() ? s.next() : "";
+   }
+
+   /**
+    * @see javax.ws.rs.core.Application#getClasses()
+    */
+   @Override
+   public Set<Class<?>> getClasses() {
+      if (classes.isEmpty()) {
+         for (String clazz : getClassesFromDeployment("classes")) {
```

Comment:
Can you please clarify me a bit what you're doing in this Application class and why? In particular, I'm concerned with you reading from the classes.txt file, as AFAIR that's something our TestUtil class adds to deployment just for our testsuite. Is this a shortcut or something? What would be the other approach for achieving the same if that's the case? Maybe I'm simply missing something obvious here, though.

Reply:
It seems our current testsuite uses `classes.txt` and `singleton.txt` inside the bundled war file to load them properly.

I copied it from `org.jboss.resteasy.utils.TestApplication`, and it's written by `tterem`:

```bash
$ git blame ./arquillian-utils/src/main/java/org/jboss/resteasy/utils/TestApplication.java | grep classes
48ac2ebcd3 (tterem          2018-09-25 12:32:30 +0200 18)    public static Set<Class<?>> classes = new HashSet<Class<?>>();
48ac2ebcd3 (tterem          2018-09-25 12:32:30 +0200 22)     * Load resources from classes.txt file from deployment
48ac2ebcd3 (tterem          2018-09-25 12:32:30 +0200 64)       if (classes.isEmpty()) {
48ac2ebcd3 (tterem          2018-09-25 12:32:30 +0200 65)          for (String clazz : getClassesFromDeployment("classes")) {
48ac2ebcd3 (tterem          2018-09-25 12:32:30 +0200 68)                   classes.add(Class.forName(clazz));
48ac2ebcd3 (tterem          2018-09-25 12:32:30 +0200 75)       return classes;
```

Our current tests are mainly based on above scheme, but I need to create my own singleton instance of `WadlDefaultResource` for testing, so I have to create `WadlTestApplication` to override the behavior in `TestApplication`.

Reply:
OK, how about extending TestApplication and calling super.getSingletons() in the overridden getSingletons() method?

Reply:
We can't extend `TestApplication`, because it's marked as `@Provider` and`@ApplicationPath("/")`. The extended class won't work properly:

```
Failed tests:
  DeploymentTest.testBasic:65 expected:<200> but was:<404>
```

Reply:
I also tried to create an abstract class `BasicApplication extends Application`, and then make `TestApplication` and `WadlTestApplication` to extend it, but it also not work properly. It should be relative to the way `Application` is processed.